### PR TITLE
impl(generator): update EndpointOption with result of DetermineServiceEndpoint

### LINKED ISSUE #13233: Update MakeConnection and stub creation functions to use Options modified by stub creation functions instead of just a copy of the Options provided (gRPC).

### PR DESCRIPTION
**Pull request by scotthart**
fixes #13233

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13234)

### CODE REVIEW

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13234)* status: 0 of 593 files reviewed, 1 unresolved discussion (waiting on @devbww)

*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13234#gh-1417674576:-Nl4WODCEYRz-_S1j0vQ:bg4n4t4) (no related file):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

For what it's worth, this doesn't "feel right" to me.  `Options` are things we should pass down from the user verbatim, not how we should return values from internal functions.
</blockquote></details>

If the user specified the EndpointOption to begin with, DetermineServiceEndpoint returns that same value and this is essentially a no-op. If the user specifies the env var then EndpointOption gets set to that value similar to how PopulateCommonOptions currently behaves (all EndpointOption functionality will eventually be removed from PopulateCommonOptions and performed here).
If the user specifies the UniverseDomainOption then EndpointOption gets set to the result of DetermineServiceEndpoint.
The main reason we're mutating the options is to preserve the existing behavior where the user can retrieve the endpoint being used by interrogating the Options from FooConnection::options.

___

**Review by devbww — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 13 of 592 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13234)* status: 14 of 593 files reviewed, 1 unresolved discussion (waiting on @scotthart)

___
*[`generator/internal/stub_factory_generator.cc` line 115 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13234#gh-1417674576:-Nl4kaANBxUkrRt4GXBA:bn5l0ys) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/12ebc6b293efe3ea44886a6c91aacfac9114300f/generator/internal/stub_factory_generator.cc#L115)):*
> we're mutating the options ... to preserve the existing behavior where the user can retrieve the endpoint being used by interrogating the Options from FooConnection::options.

I'm not sure we ever made that guarantee (and I'm not sure it is always true), and it is what I'm questioning.  If we want to retrieve the endpoint being used by a connection, I reckon we should ask the connection, not look at the options used to create it.

**Comment by dbolduc** [comment]
> I'm not sure we ever made that guarantee

Hyrum's Law. I think we should try to preserve the behavior of `Connection::options()`, namely that `EndpointOption` gets set to the right value.

> `Options` are things we should pass down from the user verbatim

We modify `Options` immediately to include defaults... I think of this PR as just defaulting the `Options` in a different place than we have been doing.

File: generator/integration_tests/tests/golden_kitchen_sink_stub_factory_test.cc [hunk 1]
```diff
@@ -41,7 +41,9 @@ class GoldenKitchenSinkStubFactoryTest : public ::testing::Test {
 };
 
 TEST_F(GoldenKitchenSinkStubFactoryTest, DefaultStubWithoutLogging) {
-  auto default_stub = CreateDefaultGoldenKitchenSinkStub(CompletionQueue{}, {});
```
Comment (dbolduc) [inline generator/integration_tests/tests/golden_kitchen_sink_stub_factory_test.cc:44]: archaeology nit: this doesn't belong in the "regenerate code" commit

File: generator/integration_tests/tests/golden_kitchen_sink_stub_factory_test.cc [hunk 2]
```diff
@@ -41,7 +41,9 @@ class GoldenKitchenSinkStubFactoryTest : public ::testing::Test {
...
+  Options options;
```
Comment (dbolduc) [inline generator/integration_tests/tests/golden_kitchen_sink_stub_factory_test.cc:44]: can we set expectations on `options.get<EndpointOption>()` on the other side of the `CreateDefault*Stub` call?

I know https://github.com/scotthart/google-cloud-cpp/blob/main/generator/integration_tests/tests/universe_domain_test.cc is a thing... just seems like we should have a unit test here, too.

**Review by dbolduc — APPROVED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13234)* status: 1 of 593 files reviewed, 3 unresolved discussions (waiting on @dbolduc and @devbww)

___
*[`generator/integration_tests/tests/golden_kitchen_sink_stub_factory_test.cc` line 0 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13234#gh-1423090044:-NlUJ4IPCJK_ecgzO2qT:bt2vr1b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08ebc5a1467659d7fa64b3adbf63a121cf1146be/generator/integration_tests/tests/golden_kitchen_sink_stub_factory_test.cc#L0)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/googleapis/google-cloud-cpp/pull/13234#discussion_r1423090044) to the right spot — sorry!)*

archaeology nit: this doesn't belong in the "regenerate code" commit
</blockquote></details>

Fixed.
___
*[`generator/integration_tests/tests/golden_kitchen_sink_stub_factory_test.cc` line 44 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13234#gh-1423094458:-NlUHl4d0X2xsDLkndMw:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/08ebc5a1467659d7fa64b3adbf63a121cf1146be/generator/integration_tests/tests/golden_kitchen_sink_stub_factory_test.cc#L44)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

can we set expectations on `options.get<EndpointOption>()` on the other side of the `CreateDefault*Stub` call?

I know https://github.com/scotthart/google-cloud-cpp/blob/main/generator/integration_tests/tests/universe_domain_test.cc is a thing... just seems like we should have a unit test here, too.
</blockquote></details>

Done.
___
*[`generator/internal/stub_factory_generator.cc` line 115 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13234#gh-1417674576:-NlUHtSGDYltNq1t1WQL:b-ifujnu) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/12ebc6b293efe3ea44886a6c91aacfac9114300f/generator/internal/stub_factory_generator.cc#L115)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

> we're mutating the options ... to preserve the existing behavior where the user can retrieve the endpoint being used by interrogating the Options from FooConnection::options.

I'm not sure we ever made that guarantee (and I'm not sure it is always true), and it is what I'm questioning.  If we want to retrieve the endpoint being used by a connection, I reckon we should ask the connection, not look at the options used to create it.
</blockquote></details>

Sounds like you are proposing a new member function on the Connection class to retrieve the endpoint actually used. I think this would also require additional work during stub creation or a new member function on the stub class. That may be something we want to add. These changes are meant to preserve existing behavior; we can always change to new behavior if we desire.

**Pull request merged**